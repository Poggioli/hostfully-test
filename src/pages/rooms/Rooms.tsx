import { LoadingRoomCard } from "@/components/RoomCard";
import { RoomCard } from "@/components/RoomCard/RoomCard";
import { Button } from "@/components/ui/button";
import { useStore } from "@/lib/store";
import { useGetRooms } from "@/service/Room";
import { RefreshCcw } from "lucide-react";
import { FC } from "react";

type RoomsErrorProps = {
  refetch: () => unknown
}
const RoomsError: FC<RoomsErrorProps> = ({ refetch }) => {
  return (
    <div className="flex flex-col gap-4">
      <h1>Ops, something went wrong...</h1>
      <Button onClick={refetch}>
        <RefreshCcw className="mr-2 h-4 w-4" /> Try again
      </Button>
    </div>
  )
}

const Rooms: FC = () => {

  const { rooms } = useStore()
  const { isLoading, isError, refetch } = useGetRooms();

  return (
    <div className="w-full grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-x-6 gap-y-10">
      {!isLoading && !isError && rooms.length ? (
        <>
          {rooms.map((room) => <RoomCard key={room.id} {...room} />)}
        </>
      ) : null}
      {isLoading ? (
        <>
          {Array(15).fill(0).map((_, index) => <LoadingRoomCard key={index} />)}
        </>
      ) : null}
      {isError ? (
        <RoomsError refetch={() => refetch()} />
      ) : null}
    </div>)
}

export { Rooms };
