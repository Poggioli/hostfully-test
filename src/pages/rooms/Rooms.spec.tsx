import { act, fireEvent, render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Rooms } from "./Rooms";

const mockRooms = jest.fn().mockImplementation(() => [
  {
    id: 'id-room-1',
    name: 'name rom 1',
    description: 'description room 1',
    pricePerDay: 500,
    pricePerGuest: 100,
    photos: ['', ''],
  }
])

jest.mock("@/lib/store", () => {
  return {
    useStore: () => ({
      bookings: [],
      rooms: (() => mockRooms())()
    }),
  };
});

const mockIsLoading = jest.fn().mockReturnValue(false);
const mockIsError = jest.fn().mockReturnValue(false);
const mockRefetch = jest.fn();
jest.mock("@/service/Room", () => ({
  useGetRooms: () => ({
    isLoading: (() => mockIsLoading())(),
    isError: (() => mockIsError())(),
    refetch: mockRefetch
  }),
}));

describe("Rooms", () => {
  it(`GIVEN a Rooms component
      WHEN the useGetRooms return success
      AND the data has any value
      THEN should render the rooms card`, () => {
    const { baseElement } = render(
      <BrowserRouter>
        <Rooms />
      </BrowserRouter>
    );
    expect(baseElement).toMatchSnapshot();
  })

  it(`GIVEN a Rooms component
      WHEN the useGetRooms return success
      AND the data is empty
      THEN should render the empty message`, () => {
    mockRooms.mockReturnValueOnce([]);
    const { baseElement } = render(
      <BrowserRouter>
        <Rooms />
      </BrowserRouter>
    );
    expect(baseElement).toMatchSnapshot();
  })

  it(`GIVEN a Rooms component
      WHEN the useGetRooms is loading
      THEN should render the loading card`, () => {
    mockIsLoading.mockReturnValueOnce(true);
    const { baseElement } = render(
      <BrowserRouter>
        <Rooms />
      </BrowserRouter>
    );
    expect(baseElement).toMatchSnapshot();
  })

  it(`GIVEN a Rooms component
      WHEN the useGetRooms return error
      THEN should render the error message`, () => {
    mockIsError.mockReturnValueOnce(true);
    const { baseElement } = render(
      <BrowserRouter>
        <Rooms />
      </BrowserRouter>
    );
    expect(baseElement).toMatchSnapshot();
  })

  it(`GIVEN useGetRooms with error
      WHEN user click in "Try again" button
      THEN should refetch`, () => {
    mockIsError.mockReturnValueOnce(true);
    const { getByRole } = render(
      <BrowserRouter>
        <Rooms />
      </BrowserRouter>
    );
    const tryAgainButton = getByRole('button', {
      name: 'Try again'
    });

    act(() => fireEvent.click(tryAgainButton));

    expect(mockRefetch).toHaveBeenCalledTimes(1);
  })
})