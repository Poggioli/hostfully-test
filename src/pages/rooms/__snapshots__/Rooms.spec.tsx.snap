// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms GIVEN a Rooms component
      WHEN the useGetRooms is loading
      THEN should render the loading card 1`] = `
<body>
  <div>
    <div
      class="w-full grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-x-6 gap-y-10"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-[200px] w-full"
        />
        <div
          class="p-3 flex flex-col gap-4"
        >
          <div
            class="animate-pulse rounded-md bg-primary/10 h-6 w-full max-w-[150px]"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-14 w-full"
          />
          <div
            class="animate-pulse rounded-md bg-primary/10 h-5 w-full max-w-[100px]"
          />
          <div
            class="flex justify-between gap-4"
          >
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
            <div
              class="animate-pulse rounded-md bg-primary/10 h-9 w-full max-w-[100px]"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Rooms GIVEN a Rooms component
      WHEN the useGetRooms return error
      THEN should render the error message 1`] = `
<body>
  <div>
    <div
      class="w-full flex flex-col justify-center gap-4 items-center"
    >
      <h1>
        Ops, something went wrong...
      </h1>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
      >
        <svg
          class="lucide lucide-refresh-ccw mr-2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 12a9 9 0 0 0-9-9 9.75 9.75 0 0 0-6.74 2.74L3 8"
          />
          <path
            d="M3 3v5h5"
          />
          <path
            d="M3 12a9 9 0 0 0 9 9 9.75 9.75 0 0 0 6.74-2.74L21 16"
          />
          <path
            d="M16 16h5v5"
          />
        </svg>
         Try again
      </button>
    </div>
  </div>
</body>
`;

exports[`Rooms GIVEN a Rooms component
      WHEN the useGetRooms return success
      AND the data has any value
      THEN should render the rooms card 1`] = `
<body>
  <div>
    <div
      class="w-full grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-x-6 gap-y-10"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow w-full"
      >
        <div
          data-radix-aspect-ratio-wrapper=""
          style="position: relative; width: 100%; padding-bottom: 56.25%;"
        >
          <div
            class="rounded-t-md overflow-hidden"
            style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
          >
            <img
              alt="Photo of name rom 1"
              class="rounded object-cover w-full h-full"
              loading="lazy"
              src=""
            />
          </div>
        </div>
        <div
          class="space-y-1.5 p-6 flex flex-col gap-4 pb-3"
        >
          <h3
            class="font-semibold leading-none tracking-tight"
          >
            name rom 1
          </h3>
          <p
            class="text-sm text-muted-foreground line-clamp-3"
          >
            description room 1
          </p>
        </div>
        <div
          class="p-6 pt-0 flex justify-end pb-4"
        >
          <span>
            <span
              class="text-sm font-bold"
            >
              $
              500
            </span>
             
            night
          </span>
        </div>
        <div
          class="items-center p-6 pt-0 flex justify-end"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
          >
            Reserve
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Rooms GIVEN a Rooms component
      WHEN the useGetRooms return success
      AND the data is empty
      THEN should render the empty message 1`] = `
<body>
  <div>
    <div
      class="w-full grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-x-6 gap-y-10"
    >
      <h1>
        Ops, we don't have rooms available at this moment...
      </h1>
    </div>
  </div>
</body>
`;
